Refactor AutoStringRefAs* classes

peter-b opened the pull request:
When compiling LiveCode using GCC6, the compiler was getting very
upset by several strict aliasing violations in
`MCAutoStringRefAsNativeChars`.  Shutting it up was a good excuse for
a refactor.  Accordingly:
- No strict-aliasing warnings from GCC 6
- The underlying stringref is cached in an MCAutoStringRef, so no
  explicit dereferencing or destruction is needed
- No messing about with buffers, by using `MCStringNativeCopy()` to
  obtain a nativized string without modifying the original string
- `MCStringRefAsNativChars` has an API that's more compatible with the
  other `MCAutoStringRefAs*` classes, including a single-operand
  `Lock()`, a `Size()` method and an `operator*` implementation
- `MCStringRefAsCString` is a small subclass of
  `MCStringRefAsNativeChars` that dereferences as a `char *` instead
  of a `char_t *`

peter-b committed:
libfoundation: Refactor AutoStringRefAs{C,Native} classes.

When compiling LiveCode using GCC6, the compiler was getting very
upset by several strict aliasing violations in
`MCAutoStringRefAsNativeChars`.  Shutting it up was a good excuse for
a refactor.  Accordingly:

- No strict-aliasing warnings from GCC 6
- The underlying stringref is cached in an MCAutoStringRef, so no
  explicit dereferencing or destruction is needed
- No messing about with buffers, by using `MCStringNativeCopy()` to
  obtain a nativized string without modifying the original string
- `MCStringRefAsNativChars` has an API that's more compatible with the
  other `MCAutoStringRefAs*` classes, including a single-operand
  `Lock()`, a `Size()` method and an `operator*` implementation
- `MCStringRefAsCString` is a small subclass of
  `MCStringRefAsNativeChars` that dereferences as a `char *` instead
  of a `char_t *`

livecodeali commented:
@livecode-vulcan review ok f2cdd2d

livecode-vulcan commented:
💙 review by @livecodeali ok f2cdd2d4248d6c406f2cee37252883c850c054c0

livecode-vulcan commented:
😞 test failure f2cdd2d4248d6c406f2cee37252883c850c054c0
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2780)
- try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/2789)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2775)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2791)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/2797)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2801)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2819)

peter-b committed:
MCS_startprocess_unix: Don't mutate MCAutoStringRefAsUTF8String

The internal buffer of an `MCAutoStringRefAsUTF8String` should be
treated as constant by the code that uses it.

This patch modifies `MCS_startprocess_unix()` to call
`MCStringConvertToUTF8String()` directly and drop the allocated
buffer without cleaning it up, since the subsequent `execve()` will
nuke everything anyway.

peter-b commented:
@livecodeali I think I've fixed the issue that caused the Mac build failure.

livecodeali commented:
I wonder why the argument parsing was never stringref-ified. I guess it's unlikely that people will be using non-ascii space characters in their process argument strings anyway :-)
@livecode-vulcan review ok a51231d

livecode-vulcan commented:
💙 review by @livecodeali ok a51231dd73983be0e6a23b3050be34684fe88eed

livecode-vulcan commented:
😞 test failure a51231dd73983be0e6a23b3050be34684fe88eed
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2786)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2795)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2781)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2797)
- try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/2803)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2807)
- try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2825)

peter-b committed:
apk_list_folder_entries(): Don't mutate MCAutoStringRefAsUTF8String

Since `apk_list_folder_entries()` needs a mutable UTF-8 string buffer,
explicitly create an owned buffer rather than mutating the internal
buffer of a `MCAutoStringRefAsUTF8String`.

peter-b commented:
@livecodeali And now I think I've managed to deal with the Android-specific issues...

livecodeali commented:
@livecode-vulcan review ok 5f4790f

livecode-vulcan commented:
💙 review by @livecodeali ok 5f4790f2512a35839e32f2095ace305bc98eed4b

livecode-vulcan commented:
😞 test failure 5f4790f2512a35839e32f2095ace305bc98eed4b
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2792)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2801)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2787)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2803)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2809)
- try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2813)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2832)

peter-b committed:
libfoundation: Dereference AutoStringRefAs* instances as const.

The buffers that the `AutoStringRefAs*` utility classes expose are not
mutable, and dereferencing them with `operator*` shouldn't mutate the
underlying class.

There's one exception: `MCAutoStringRefAsWString` is quite often used
to provide a `LPWSTR` for passing to Windows API functions.  Many of
these `LPWSTR` arguments are _not_ `const`, and having to `const_cast`
many uses of `MCAutoStringRefAsWString` would remove a lot of the
utility of the class.

peter-b committed:
libfoundation: Refactor MCAutoStringRefAsWString.

`MCAutoStringRefAsWString` was previously used:

- where a string is needed as a pointer to a UTF-16 string buffer;
  this is covered by the `MCAutoStringRefAsUTF16String` class

- where a string is needed as an `LPWSTR` for passing to Win32 API
  functions; this is covered by the new `MCAutoStringRefAsLPWSTR`
  class which is only defined on Windows

This patch makes `MCAutoStringRefAsWString` and
`MCAutoStringRefAsLPCWSTR` into synonyms for
`MCAutoStringRefAsUTF16String`.

It also adds a `MCAutoStringRefAsLPWSTR` class.  Since this regularly
used to pass a mutable `LPWSTR` (even though usually the buffer isn't
expected to be modified), it always does a deep copy.

peter-b committed:
libfoundation: Add MCStringUnicodeCopy().

The opposite of `MCStringNativeCopy()`.  As a side-effect, fixes a
failure-path memory leak in `MCStringNativeCopy()`.

peter-b committed:
Use MCAutoStringRefAsLPWSTR where mutable buffer may be required

Some Win32 APIs take a `LPWSTR` rather than a `LPCWSTR` argument.
Ensure that if (for some reason) they try to mutate the buffer that
they're passed, it doesn't affect the original LiveCode string.

livecodeali commented:
@livecode-vulcan review ok d970d9f

livecode-vulcan commented:
💙 review by @livecodeali ok d970d9f15d9c543ff43bbf8cedb9d4d93b8289ec

livecode-vulcan commented:
😎 test success d970d9f15d9c543ff43bbf8cedb9d4d93b8289ec
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2812)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2819)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2805)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2821)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2827)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2836)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2850)
